### ISSUE TITLE
Question: why object type is mapped as interface{} and not map[string]interface{}?

### ISSUE THREAD

**Issue by fredbi** [go-swagger/go-swagger#1450]
## Problem statement
I would like what is the reason we map an untyped json object as interface{} rather than map[string]interface{} in models.
The generator understands this is a map but still, it generates an interface.
Take the Foo property in the example below.
It has the following generation attributes:
- IsAnonymous: true (why?)
- IsAliased: true
- IsMap: true
- IsInterface: true

This led to an uncovered corner case for validation that I fixed with PR #1448.
Nonetheless, I feel like the "object" specification should reject input such as a slice.

Please advise.

Example (from issue #910):
swagger: "2.0"

info:
  title: Test
  description: Test
  version: "1.0.0"

paths:
  /test:
    get:
      description: test
      responses:
        200:
          schema:
            type: object
            properties:
              foo:
                type: object
              bar:
                type: integer
                format: int64
              baz:
                type: integer
                format: int64
              quux:
                type: array
                items:
                  type: string
            required:
              - foo
              - bar

## Environment
swagger version: x.x.x  
go version: x.x.x  
OS:

**Comment by gregmarr** [comment]
I just ran into this myself.  I think this would actually be fixed by properly handling the default of `additionalProperties` being `true` everywhere, because this should really be a struct containing the `additionalProperties` map, because anything in here would fall under `additionalProperties`.

**Comment by gregmarr** [comment]
Unfortunately, adding `additionalProperties: true` to the schema doesn't do anything.

**Comment by fredbi** [comment]
> Unfortunately, adding additionalProperties: true to the schema doesn't do anything.
This could just be a matter of setting some default behavior. The current mapping strategy assumes that undefined additional properties won't be consumes and favors a tighter model against a little breach in the JSON-schema. I don't need this one now, but adding such a generation option is quite doable.

My point lies somewhere else: I wouldn' have expected that an object correctly identified as a Map by the analyzer is rendered as interface{}.

I admit I tagged this additionalProperties maybe a bit too lightheartedly... In my (prehaps uninformed) view, map[] <=> additionalProps,...

**Comment by gregmarr** [comment]
I was hoping that correct handling of `additionalProperties` would have turned it into a struct with an additionalProperties member, and nothing special would be needed for this case, but I guess there's an override somewhere for a type with no explicit properties.

**Comment by fredbi** [comment]
Yes. That's precisely why I am asking for any special reason to prefer interface{} to map[string]interface{}.
I might have missed something...

**Comment by gregmarr** [comment]
Well, only that the proper way to do this, to match all the other types, is to have an empty struct, and to have the `additionalProperties` member if it is enabled.

**Comment by casualjim** [comment]
the short answer is there is no reason. I took `type: object` to mean java.lang.Object meaning it can be anything and as such I went with `interface{}`

**Comment by fredbi** [comment]
I think that the correct behavior should be:
- empty schema (`{}`) => `interface{}`
- empty object => `map[string]interface{}` (that works for instance when you state: `"additionalProperties": {}`
- empty items in array (forbidden by swagger, but supported in codegen): `[]interface{]`

It works as expected like this in most instances but in the case of empty object with `"additionalProperties"` let unspecified.

Obviously, this is a minor finicky point...

**Comment by fredbi** [comment]
updated NOTE: there has been a slight improvement on the latest master. Now an object with no schema, but with validation "minProperties" or "maxProperties" is correctly inferred as a `map[string]interface{}` not just `interface{}`
